feat: migration from 2.2 -> 3.2 -> 4.2

JacobCoffee opened the pull request:
## What

- Migration of Django from 2.x to 4.2

## Closes

- Fixes #1860

JacobCoffee committed:
feat: initial 2.2 -> 4.2 migration work

JacobCoffee committed:
feat: minimum dependency changes required to get a live webserver:port log line

JacobCoffee committed:
test(minutes): fixed minutes test

JacobCoffee committed:
test(mailing): fixed mailing test

JacobCoffee committed:
test(codesamples): fixed codesamples tests

JacobCoffee committed:
test(downloads): fixed downloads tests

note:  unsure if "valid"

JacobCoffee committed:
test(pages): fixed pages tests

JacobCoffee committed:
test(successstories): fixed successstories tests

JacobCoffee committed:
test(users): fixed users forms tests

JacobCoffee committed:
test(events): fixed events tests

JacobCoffee committed:
test(users): fixed users views tests

JacobCoffee committed:
chore: hide my shame

JacobCoffee committed:
chore: apply code review

ewdurbin committed:
harmonize docker-compose and ci.yml PG versions to prod

JacobCoffee committed:
Update base-requirements.txt

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

JacobCoffee committed:
chore: remove migrations, rebase latest

JacobCoffee committed:
chore: apply code review

JacobCoffee committed:
feat(db): generate migrations

ewdurbin commented on base-requirements.txt:
```diff
@@ -22,32 +22,31 @@ chardet==4.0.0
+django-markupfield==2.0.1
 
-django-allauth==0.50.0
+django-allauth==0.61.1  # 0.55.0 is first version that supports Django 4.2
```
```suggestion
django-allauth==0.57.2  # 0.55.0 is first version that supports Django 4.2
```

[This migration from django-allauth version 0.58](https://github.com/pennersr/django-allauth/blob/main/allauth/socialaccount/migrations/0006_alter_socialaccount_extra_data.py) was unable to apply to a copy of the existing production database due to:

```
  Applying socialaccount.0006_alter_socialaccount_extra_data...Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 87, in _execute
    return self.cursor.execute(sql)
psycopg2.errors.DatatypeMismatch: default for column "extra_data" cannot be cast automatically to type jsonb


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/./manage.py", line 22, in <module>
    main()
  File "/code/./manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 356, in handle
    post_migrate_state = executor.migrate(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/operations/fields.py", line 235, in database_forwards
    schema_editor.alter_field(from_model, from_field, to_field)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 831, in alter_field
    self._alter_field(
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/schema.py", line 288, in _alter_field
    super()._alter_field(
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 1056, in _alter_field
    self.execute(
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/schema.py", line 48, in execute
    return super().execute(sql, None)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 201, in execute
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 102, in execute
    return super().execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 87, in _execute
    return self.cursor.execute(sql)
django.db.utils.ProgrammingError: default for column "extra_data" cannot be cast automatically to type jsonb
```

We should probably hold off on this dependency upgrade temporarily until that can be addressed separately since support for 4.2 is available in the release right before it.

JacobCoffee replied:
Loading at 0.55.0 and applying migrations worked happily, although CI is now failing on some tests. Will investigate... iirc this is why I went above the minimum package version that had 4.2 support
```
======================================================================
ERROR: test_duplicate_email (users.tests.test_forms.UsersFormsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/code/users/tests/test_forms.py", line 65, in test_duplicate_email
    form.save(request)
  File "/usr/local/lib/python3.9/site-packages/allauth/account/forms.py", line 444, in save
    setup_user_email(request, user, [])
  File "/usr/local/lib/python3.9/site-packages/allauth/account/utils.py", line 273, in setup_user_email
    stashed_email = adapter.unstash_verified_email(request)
  File "/usr/local/lib/python3.9/site-packages/allauth/account/adapter.py", line 65, in unstash_verified_email
    ret = request.session.get("account_verified_email")
AttributeError: 'WSGIRequest' object has no attribute 'session'

======================================================================
FAIL: test_user_duplicate_username_email (users.tests.test_views.UsersViewsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/test/utils.py", line 461, in inner
    return func(*args, **kwargs)
  File "/code/users/tests/test_views.py", line 246, in test_user_duplicate_username_email
    self.assertContains(
  File "/usr/local/lib/python3.9/site-packages/django/test/testcases.py", line 660, in assertContains
    self.assertTrue(
AssertionError: False is not true : Couldn't find 'A user is already registered with this email address.' in response

----------------------------------------------------------------------
Ran 621 tests in 51.930s

FAILED (failures=1, errors=1)
Destroying test database for alias 'default'...
make: *** [test] Error 1
```

JacobCoffee replied:
If I create 2 users, `newuser` and `existinguser`, with the same email it allows me to do this in this version. It only is concerned with unique usernames.

JacobCoffee replied:
This is not a problem in later versions, but if needed we could maybe override `SignupForm` but that doesnt feel great

ewdurbin replied:
In my testing `0.57.2` was working well. does that exhibit the issue?

JacobCoffee replied:
I'll try again 🙃

JacobCoffee replied:
Just an error with that version (not failure) 
```
======================================================================
ERROR: test_duplicate_email (users.tests.test_forms.UsersFormsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/code/users/tests/test_forms.py", line 65, in test_duplicate_email
    form.save(request)
  File "/usr/local/lib/python3.9/site-packages/allauth/account/forms.py", line 447, in save
    setup_user_email(request, user, [])
  File "/usr/local/lib/python3.9/site-packages/allauth/account/utils.py", line 337, in setup_user_email
    stashed_email = adapter.unstash_verified_email(request)
  File "/usr/local/lib/python3.9/site-packages/allauth/account/adapter.py", line 69, in unstash_verified_email
    ret = request.session.get("account_verified_email")
AttributeError: 'WSGIRequest' object has no attribute 'session'
```

ewdurbin replied:
Resolve, See b3bd4969f04808dcaec6793491247c138c4909a4

JacobCoffee committed:
fix: revert allauth version bump

Resolves https://github.com/python/pythondotorg/pull/2520#discussion_r1747192645

JacobCoffee committed:
fix: revert allauth account middlware used in later versions

JacobCoffee committed:
chore: bump allauth version

ewdurbin committed:
Fixup tests that interact with django-allauth ACCOUNT_PREVENT_ENUMERATION

ACCOUNT_PREVENT_ENUMERATION was introduced in django-allauth 0.52.0, and interferes with our expectations.

This should probably be turned on! But for now disable it by default to keep the changeset minimal.

Allauth _used_ to iterate over users to check for email uniquenss but stopped at some point, so we have to create an EmailAdress object for the user in the relevant test-case for duplicate emails

JacobCoffee committed:
Merge branch 'main' into 4.2

JacobCoffee committed:
chore: bump django package version
